### PR TITLE
ExpiringMultipleLevelCache: Caching Error

### PR DESCRIPTION
**Pull request by leonardonsantos**
Adding option to cache Errors, with different TTL

This ExpiringMultipleLevelCache is used for cache some external request. When those requests found a status 404 or even a status 500 they throw exception. So I'm adding the ability to cache those errors, even if they have little TTL at cache. This means that during TTL of those cached errors, the request won't be made again.

### CODE REVIEW

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala [hunk 1]
```diff
@@ -17,9 +17,23 @@ import scala.util.control.NonFatal
...
+                                 hasError: Boolean = false,
+                                 throwable: Option[Throwable] = None) {
+    def hasExpired(ttl: FiniteDuration, now: DateTime, ttlCachedErrors: FiniteDuration = 1.minute): Boolean = {
+      if (!hasError) {
+        date.plus(ttl.toMillis).isBefore(now)
+      } else {
+        date.plus(ttlCachedErrors.toMillis).isBefore(now)
+      }
+    }
+    def getValue: V = {
```
Comment (fernandors87) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: Why are you converting an Option to Exception?
Just use it as an Option.

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala [hunk 2]
```diff
@@ -128,94 +143,117 @@ case class ExpiringMultiLevelCache[V](ttl: FiniteDuration,
...
+
+  private def timestamp(hasError: Boolean, throwable: Throwable): TimestampedValue[V] = {
+    TimestampedValue(value = None, date = now, hasError = hasError, throwable = Some(throwable))
+  }
 
   private def elapsedTime(startNanoTime: Long) = FiniteDuration(System.nanoTime() - startNanoTime, TimeUnit.NANOSECONDS)
 
   private def remoteLockKey(key: Any) = s"$key-emlc-lock"
 
+  private def checkSavedErrorCache(key: String, genValue: () => Future[V], startTime: Long, v: TimestampedValue[V]): Future[V] = {
```
Comment (fernandors87) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: Why this method exists?
The only thing it does is convert an Option[Throwable] in a Future[V].

```scala
v.throwable.map(Future.failed[V]).orElse(Future.successful(v.value))
```

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala [hunk 3]
```diff
@@ -17,9 +17,23 @@ import scala.util.control.NonFatal
...
+        date.plus(ttl.toMillis).isBefore(now)
+      } else {
+        date.plus(ttlCachedErrors.toMillis).isBefore(now)
+      }
+    }
+    def getValue: V = {
+      this.value match {
+        case Some(x) => x
+        // We should never try to get a value that was saved as None, probably from an error
+        case None => throw new Exception("Trying to get None value")
```
Comment (fernandors87) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: Remember: unhandled exceptions and nulls are your enemies!!

If you **REALLY** want this exception, make sure it's wrapped by a Try or a Future

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala (hunk 1 above)
Comment (fernandors87) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: You already have value property exposed. You don't need this method.

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala [hunk 4]
```diff
@@ -370,7 +411,18 @@ case class ExpiringMultiLevelCache[V](ttl: FiniteDuration,
           case None =>
             // There are no remote RW caches
             logger.error(s"canonicalValueGenerator, key $key: failed to generate value and no remote cache configured", eLocal)
-            Future.failed(eLocal)
+            eLocal match {
+              case NonFatal(e) => {
+                if (cacheErrors) {
+                  // if error was NonFatal Error then saves it to cache
+                  val timestampedValue = timestamp(hasError = true, throwable = e)
```
Comment (fernandors87) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: I think the throwable is good enough to say it has an error or not.
I'd remove the hasError flag. It's redundant

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala [hunk 5]
```diff
@@ -17,9 +17,23 @@ import scala.util.control.NonFatal
 import scala.util.{Failure, Success, Try}
 
 object ExpiringMultiLevelCache {
-  case class TimestampedValue[V](date: DateTime, value: V) {
-    def hasExpired(ttl: FiniteDuration, now: DateTime): Boolean = {
-      date.plus(ttl.toMillis).isBefore(now)
+  case class TimestampedValue[V](date: DateTime,
+                                 value: Option[V] = None,
+                                 hasError: Boolean = false,
+                                 throwable: Option[Throwable] = None) {
```
Comment (fernandors87) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: I'd change this class constructor to:

```scala
case class TimestampedValue[V](date: DateTime, value: Try[V])
```

- I think you don't need the hasError flag.
- Try[V] can be a actual value or a throwable, so you don't need the extra "throwable" property.

Comment (leonardonsantos) [reply src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: @fernandors87 All those things you pointed out can be solved with this `Try[V]`. Thank you.

**Comment by leonardonsantos** [comment]
@fernandors87 Can you take a look again?

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala [hunk 6]
```diff
@@ -141,81 +148,81 @@ case class ExpiringMultiLevelCache[V](ttl: FiniteDuration,
...
+          case Success(localValue) if !localValue.hasExpired(ttl, now, ttlCachedErrors) =>
             // We have locally a good value, just return it
             reporter.onLocalCacheHit(key, elapsedTime(startTime))
             // But if we're paranoid, let's check if the local value is consistent with remote
             if (sanityLocalValueCheck)
-              remoteRW.map(remote => sanityLocalValueCheck(key, localValue, remote, genValue, startTime)).getOrElse(Future.successful(localValue.value))
+              remoteRW.map(remote => sanityLocalValueCheck(key, localValue, remote, genValue, startTime)).getOrElse(Future.successful(localValue.value.get))
             else
-              Future.successful(localValue.value)
+              Future.successful(localValue.value.get)
```
Comment (fernandors87) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: You're not checking the localValue.value type.
If localValue.value is a failure, ".get" will throw an exception and you're wrapping it in a successful future.
When you're dealing with monad-like structures in scala, avoid the use of methods like "get". Prefer .map, .flatMap or pattern matching instead.

File: src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala [hunk 7]
```diff
@@ -141,81 +148,81 @@ case class ExpiringMultiLevelCache[V](ttl: FiniteDuration,
...
             // Expired remote, return the it, async update
             reporter.onCacheMissButFoundExpiredRemote(key, elapsedTime(startTime))
             tryGenerateAndSet(key, genValue, startTime).map(_.value)
-            Future.successful(expiredRemote.value)
+            Future.successful(expiredRemote.value.get)
           case Success(None) =>
             // No good remote, sync generate
             reporter.onCacheMissNothingFound(key, elapsedTime(startTime))
-            tryGenerateAndSet(key, genValue, startTime).map(_.value)
+            tryGenerateAndSet(key, genValue, startTime).map(_.value.get)
```
Comment (leonardonsantos) [inline src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: @fernandors87 nice catch about `Future.fromTry`.
And about this point in the code?

What do you think about this solution?

```
tryGenerateAndSet(key, genValue, startTime).flatMap(tsv => Future.fromTry(tsv.value))
```

I am not sure about this `flatMap`. Is it blocking?

Comment (fernandors87) [reply src/main/scala/ignition/core/cache/ExpiringMultiLevelCache.scala]: tryGenerateAndSet returns a Future.
All future ops are non blocking.

Another way to write the same code:

```scala
for {
  tsv <- tryGenerateAndSet(key, genValue, startTime)
  value <- Future.fromTry(tsv.value)
} yield value

**Comment by leonardonsantos** [comment]
@fernandors87 I have just sent some changes about the issues on value.get, using Future.fromTry
Do you mind to take a look one more time? Thanks.

**Comment by leonardonsantos** [comment]
@fernandors87 @fparisotto I've run some test in mail-ops, but I was in first version of this code. And yes, I ensure that the value was correctly cached and the request was not made again. I'll run it again.

Is that ok to make it on mail-ops, or were you thinking about production test, with just one instance?

**Comment by hensg** [comment]
@leonardonsantos Could you please(pleaaaaseee) fix this annoying (really annoooooying test) that sometimes fails and sometimes pass? 

CircleCI log
```
[info] ExpiringMultipleLevelCacheSpec:
[info] ExpiringMultipleLevelCache
[info] - should calculate a value on cache miss and return it
[info] - should calculate a value on cache miss and return a failed future of the calculation
[info] - should calculate a value on cache miss after ttl *** FAILED ***
[info]   3 was not equal to 2 (ExpiringMultipleLevelCacheSpec.scala:58)
[info] - should calculate a value on cache miss just once, the second call should be from cache hit
[info] - should calculate a value on cache miss on every request
[info] - should calculate a value on cache miss, then wait ttlCachedError to get a cache miss again
```

**Comment by leonardonsantos** [comment]
@hensg I have never got this error. How can I reproduce it?

**Comment by hensg** [comment]
@leonardonsantos Just rebuild a `ignition-core` based project a couple of times. Maybe, we have some non-deterministic test or our `datalake-ignition` is missing some configuration.

image

**Comment by fparisotto** [comment]
@hensg @leonardonsantos just to let you know guys, this error is older than this recent change, there is nothing related with this new feature

**Comment by leonardonsantos** [comment]
@hensg I did that. I rebuilt the ignition-core project, we rebuild mail-widget-api (ignition-core based project) almost every day, locally or at CircleCI, and we have never got this error.
(There are some non deterministic tests that fail, but they make requests to some client's site)

Is it anything about java version or sbt version? I'm newbie at this :stuck_out_tongue_closed_eyes:

**Pull request merged**